experimental/vindex/map.go

```diff
@@ -131,46 +161,53 @@ func (b *VerifiableIndex) Lookup(key string) (indices []uint64) {
 // Update checks the input log for a new Checkpoint, and ensures that the Verifiable Index
 // is updated to the corresponding size.
 func (b *VerifiableIndex) Update(ctx context.Context) error {
-
-	var err error
-	cpSize, rawCp, _, err := b.log.GetLatestCheckpoint(ctx)
+	rawCp, err := b.inputLog.GetCheckpoint(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get latest checkpoint from DB: %s", err)
+	}
+	cp, err := b.logParseFn(rawCp)
 	if err != nil {
-		return fmt.Errorf("failed to get latest checkpoint from DB: %v", err)
+		return fmt.Errorf("failed to parse checkpoint: %s", err)
 	}
-	if cpSize == b.cpSize {
+
+	if cp.Size == b.cpSize {
 		klog.V(1).Infof("No update needed: checkpoint size is still %d", b.servingSize)
 		return nil
 	}
-	b.cpSize = cpSize
+	b.cpSize = cp.Size
 	b.rawCp = rawCp
 	klog.Infof("Building map to log size of %d", b.cpSize)
 
 	eg, cctx := errgroup.WithContext(ctx)
-	eg.Go(func() error { return b.syncFromDatabase(cctx) })
+	eg.Go(func() error { return b.syncFromInputLog(cctx) })
 	eg.Go(func() error { return b.buildMap(cctx) })
 
 	b.servingSize = b.cpSize
 
 	return eg.Wait()
 }
 
-// syncFromDatabase reads the latest checkpoint from the DB mirror of the log.
-// It stores this checkpoint in the builder, and ensures the WAL is updated with
-// all VIndex updates required to represent the log.
-func (b *VerifiableIndex) syncFromDatabase(ctx context.Context) error {
+// syncFromInputLog reads the latest checkpoint from the input log, and ensures that the WAL
+// contains a corresponding entry for every index committed to by that checkpoint.
+//
+// TODO(mhutchinson): this doesn't perform any validation on the input log to check the
+// leaves correspond to the checkpoint root hash. This was reasonable while it was based on the
+// cloneDB, which performed this validation. Implementing this will require the index to store some
+// state alongside the WAL which contains a compact range of its current progress.
+func (b *VerifiableIndex) syncFromInputLog(ctx context.Context) error {
 	if b.cpSize > b.nextIndex {
-		leaves := make(chan logdb.StreamResult, 1)
-		go b.log.StreamLeaves(ctx, b.nextIndex, b.cpSize, leaves)
+		leaves := make(chan InputLeaf, 1)
+		go b.inputLog.StreamLeaves(ctx, b.nextIndex, b.cpSize, leaves)
```

Comment:
Just looking here, I think it'd be good to do the iterator change you mentioned in the comment above; you'll leak goroutines (and possible DB sessions, etc) if e.g. you get a `mapErr` below or an error appending to the `wal`, as you've no way to tell the underlying impl of `StreamLeaves` that it should terminate (other than perhaps the overall `ctx` becoming `Done`).